zipline/finance/controls.py

```diff
@@ -286,12 +297,28 @@ def validate(self,
         Fail if the algo has passed this Asset's end_date, or before the
         Asset's start date.
         """
+        # If the order is for 0 shares, then silently pass through.
+        if amount == 0:
+            return
+
+        normalized_algo_dt = pd.Timestamp(algo_datetime).normalize()
+
         # Fail if the algo is before this Asset's start_date
-        if asset.start_date and (algo_datetime < asset.start_date):
-            self.fail(asset, amount, algo_datetime)
+        if asset.start_date:
+            normalized_start = pd.Timestamp(asset.start_date).normalize()
+            if normalized_algo_dt < normalized_start:
+                metadata = {
+                    'asset_start_date': normalized_start
+                }
+                self.fail(asset, amount, algo_datetime, metadata=metadata)
         # Fail if the algo has passed this Asset's end_date
-        if asset.end_date and (algo_datetime >= asset.end_date):
-            self.fail(asset, amount, algo_datetime)
+        if asset.end_date:
+            normalized_end = pd.Timestamp(asset.end_date).normalize()
+            if normalized_algo_dt > normalized_end:
```

Comment:
Yes (and the test case I added is to confirm it). The normalized end_date of "9/17/2015 00:00" should mean "This asset is still trading on the day of 9/17/2015". So an algo datetime of mid-day 9/17/2015 should be able to buy this.